lib/rubygems/core_ext/kernel_warn.rb

```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+if RUBY_VERSION >= "2.5"
+  module Kernel
+    path = "#{__dir__}/"
+    original_warn = instance_method(:warn)
+    Module.new {define_method(:warn, original_warn)}
+    original_warn = method(:warn)
+
+    module_function define_method(:warn) {|*messages, uplevel: nil|
+      if uplevel
+        uplevel, = [uplevel].pack("l!").unpack("l!")
+        if uplevel >= 0
+          start = 0
+          begin
+            loc, = caller_locations(start, 1)
+            break start += uplevel unless loc
+            start += 1
+          end while (loc.path.start_with?(path) or (uplevel -= 1) >= 0)
+          uplevel = start
+        end
+        original_warn.call(*messages, uplevel: uplevel)
+      else
+        original_warn.call(*messages)
+      end
```

Comment:
Instead of:

```ruby
if uplevel
  <...>
else
  original_warn.call(*messages)
end
```

perhaps use a guard clause like this:

```ruby
unless uplevel
  original_warn.call(*messages)
end

<...>
```

Reply:
Is it readable? I don't think so. I prefer original code-style.

Reply:
@duckinator When no `uplevel`, `original_warn` will be called twice in your code.
It is not the intended behavior.

Reply:
Sorry, I forgot a `return` in my suggestion! Looks like you figured it out, though.